### PR TITLE
4.x: Native image update

### LINKED ISSUE #7683: 4.0: native image build on quickstart-mp is not working
## Environment Details
* Helidon Version: 4.0.0-RC1
* Helidon MP
* JDK version: java 21 2023-09-19 LTS
* GraalVM Native version:
native-image 21 2023-09-19
GraalVM Runtime Environment Oracle GraalVM 21+35.1 (build 21+35-jvmci-23.1-b15)
Substrate VM Oracle GraalVM 21+35.1 (build 21+35, serial gc, compressed references)

----------

## Problem Description

```
cd examples/quickstarts/helidon-quickstart-mp
mvn install -Pnative-image
```

```
Fatal error: org.graalvm.compiler.debug.GraalError: org.graalvm.compiler.debug.GraalError: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method java.lang.Object.wait0(long) is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisFuture.setException(AnalysisFuture.java:49)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:322)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisFuture.ensureDone(AnalysisFuture.java:63)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapInstance.readFieldValue(ImageHeapInstance.java:110)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.onObjectReachable(ImageHeapScanner.java:472)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.heap.SVMImageHeapScanner.onObjectReachable(SVMImageHeapScanner.java:156)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.lambda$markReachable$5(ImageHeapScanner.java:452)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.lambda$postTask$14(ImageHeapScanner.java:695)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:187)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:171)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: org.graalvm.compiler.debug.GraalError: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method java.lang.Object.wait0(long) is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisFuture.setException(AnalysisFuture.java:49)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:322)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisFuture.ensureDone(AnalysisFuture.java:63)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.getOrCreateImageHeapConstant(ImageHeapScanner.java:212)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.heap.SVMImageHeapScanner.getOrCreateImageHeapConstant(SVMImageHeapScanner.java:109)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.createImageHeapConstant(ImageHeapScanner.java:186)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.createFieldValue(ImageHeapScanner.java:364)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.lambda$createImageHeapInstance$4(ImageHeapScanner.java:294)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	... 14 more
Caused by: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method java.lang.Object.wait0(long) is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.lookup(AnnotationSubstitutionProcessor.java:268)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:140)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisUniverse.lookupAllowUnresolved(AnalysisUniverse.java:438)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:417)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:78)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.infrastructure.UniverseMetaAccess.lookupJavaMethod(UniverseMetaAccess.java:112)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisMetaAccess.lookupJavaMethod(AnalysisMetaAccess.java:83)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.reflect.ReflectionFeature.createAccessor(ReflectionFeature.java:187)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.reflect.ReflectionFeature.getOrCreateAccessor(ReflectionFeature.java:141)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.target.ExecutableAccessorComputer.transform(ExecutableAccessorComputer.java:43)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.ComputedValueField.computeValue(ComputedValueField.java:366)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.substitute.ComputedValueField.readValue(ComputedValueField.java:336)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readHostedFieldValue(AnalysisConstantReflectionProvider.java:224)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.heap.SVMImageHeapScanner.readHostedFieldValue(SVMImageHeapScanner.java:124)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.createImageHeapInstance(ImageHeapScanner.java:287)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.createImageHeapObject(ImageHeapScanner.java:254)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.lambda$getOrCreateImageHeapConstant$2(ImageHeapScanner.java:205)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	... 21 more
----------------------------------------------------------------------------------------
        3.4s (8.0% of total time) in 152 GCs | Peak RSS: 1.05GB | CPU load: 6.03
========================================================================================
Finished generating 'helidon-quickstart-mp' in 41.7s.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  55.334 s
[INFO] Finished at: 2023-09-28T00:12:45Z
[INFO] ------------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by tomas-langer**
### Description
- Java 21
- latest version
- replace artifact id
- remove initialize-at-build-time in SE
- reintroduced packaging tests for native image

Resolves #7683 

Had to introduce a custom main class to modularized projects (e.g. to all MP archetypes), as the jar plugin requires a main class that is in the package of the module (it now adds the main class to manifest, and as an attribute of the `module-info.java` to make the modules runnable).

Merged changes for Kafka from @danielkec 

### Documentation

I have updated the documentation for MP to mention new requirements.

In addtion this may influence existing users of Helidon SE, as we used to have build time initialization for almost everything. Now we have aligned with GraalVM guidelines, and we do not do it for SE (except for classes where it is a requirements).

### CODE REVIEW

File: applications/se/pom.xml [hunk 1]
```diff
@@ -53,6 +53,11 @@
                                     <goal>compile</goal>
                                 </goals>
                                 <phase>package</phase>
+                                <!-- configuration>
+                                    <buildArgs>
```
Comment (romain-grecourt) [inline applications/se/pom.xml:57]: Why is this commented-out ?

File: docs/mp/aot.adoc [hunk 2]
```diff
@@ -31,10 +31,22 @@ application pom as a parent of your module, you can use the following steps to
...
-    GraalVM with `native-image` installed
+    GraalVM, or use a GraalVM JDK to build your project
+2. Add configuration to initialize your packages at build time (see below)
 2. Run Maven command `mvn clean package -Pnative-image`
 3. Execute the native executable created in `target` directory of your project
 
+== Initialize at build time
+Due to restrictions of our current implementation of integration with native image for CDI,
+we must initialize a subset of types at build time.
+As a result, all beans created within CDI need to be also initialized at build time.
```
Comment (romain-grecourt) [inline docs/mp/aot.adoc:42]: Too much internal details.
How about the following instead:

```suggestion
Helidon MP requires the types of all beans created within CDI to be initialized at build time.
```

File: examples/quickstarts/helidon-quickstart-se/pom.xml [hunk 3]
```diff
@@ -106,6 +106,20 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>org.graalvm.buildtools</groupId>
+                <artifactId>native-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>build-native-image</id>
+<!--                        <configuration>-->
```
Comment (romain-grecourt) [inline examples/quickstarts/helidon-quickstart-se/pom.xml:115]: Why is this commented out ?

File: applications/se/pom.xml (hunk 1 above)
Comment (tomas-langer) [reply applications/se/pom.xml:57]: I guess we could keep that in for SE, will uncomment.

File: examples/quickstarts/helidon-quickstart-se/pom.xml (hunk 3 above)
Comment (tomas-langer) [reply examples/quickstarts/helidon-quickstart-se/pom.xml:115]: I think I left in some of my intermediate work. Will check this

File: applications/se/pom.xml (hunk 1 above)
Comment (tomas-langer) [reply applications/se/pom.xml:57]: cannot uncomment, there is a bug in native image right now that causes this to fail. Removing from code (for now)

File: docs/mp/aot.adoc (hunk 2 above)
Comment (tomas-langer) [reply docs/mp/aot.adoc:42]: Reworded

**Comment by danielkec** [comment]
Checked Kafka connector and few more changes are needed tomas-langer/helidon/pull/4

**Review by tvallin — DISMISSED** [review]
Now that a `Main.java` class is generated for both SE and MP, it makes sense to have a `Main.java.mustache` under `archetype/common`. But this change can be made in another PR. See https://github.com/helidon-io/helidon/issues/7714

**Review by danielkec — APPROVED** [review]

**Review by tvallin — APPROVED** [review]

**Review by barchetta — APPROVED** [review]

**Pull request merged**